first go at adding macro support

KCErb opened the pull request:
Hi @Galadirith! 

I was playing around with how defining macros works and got this working locally (though I don't know how to really test it out, I just edited my .atom version of the package until I liked it and hand-copied the changes over to this fork :grin:).

The idea is that we create a local folder that we can put stuff like this into. After that Ajax just needs to know where the config file is and you're all set.

I relied on
http://docs.mathjax.org/en/latest/options/ThirdParty.html

and 

http://docs.mathjax.org/en/latest/tex.html#defining-tex-macros

pretty heavily to figure this out.

A couple of questions:

1. How can we ship out a template `macros.js` but not overwrite it on update?
2. Do we want to do something with `main.cofee`'s config? Such as 
  * Letting them specify their own config path (I played around with that for a while and kinda liked it).
  * Could a user put macro's in there? Sorta like the keymap? That would be cool!

Anyways, feel free to reject or give this harsh criticism, I won't be offended :)

KCErb commented on lib/mathjax-helper.coffee:
```diff
@@ -47,9 +47,15 @@ module.exports =
 configureMathJax = ->
+  # Add Local Path to Ajax's Config Path call it Local
+  packagePath =  atom.packages.getPackageDirPaths()[0] + '/markdown-preview-plus/local'
+  MathJax.Ajax.config.path["Local"] = packagePath
```
Is there a better way to get this path? I open to suggestions :smile:

Galadirith commented:
@KCErb Thats looks great. I gave some [suggestions](https://github.com/Galadirith/markdown-preview-plus/issues/11#issuecomment-68898026) to @darcamo last week how he might add defining macros, but it was a bit hacky and I really like your use of MathJax extensions, I didn't know MathJax could do that :D 

I was thinking about what would be the best way to implement macro's in Atom, and while a full javascript file like [macros.js](https://github.com/KCErb/markdown-preview-plus/blob/63d39e8e8e8ce4a6c704e34d76ef9d23721a0a87/local/macros.js) would be great for versatility (you could add more than just macro's to it) I don't think it fits in with the way Atom typically allows users to customize packages. All of the ones that I have seen use `.cson` files that are stored in `~/.atom`.

Package that load custom `.cson` files seem to use [season](https://github.com/atom/season) (I just got the name pun pasting it it :D), and there are a few examples of using it [here](https://github.com/atom/snippets/blob/master/lib/snippets.coffee#L60), [here](https://github.com/atom/snippets/blob/master/lib/snippets.coffee#L160), [here](https://github.com/atom-community/autocomplete-snippets/blob/master/lib/snippets-loader.coffee#L12) and [here](https://github.com/atom-community/autocomplete-snippets/blob/master/lib/snippets-loader.coffee#L53). You could then load the `.cson` file and actually add the macro contents directly to the main configuration ([Defining TeX macros](http://docs.mathjax.org/en/latest/tex.html#defining-tex-macros)).

The file itself could then be really simple, something like:
````json
Macros:
  RR: "{\\bf R}"
  bold: ["{\\bf #1}",1]
  ...
````

and perhaps an unambiguous name like `markdown-preview-plus.cson` (I was thinking perhaps `macros.cson`, but that could be macro's for any package. Also I thought `mathjax.cson` but that might be confusing to anyone who doesn't know MPP uses mathjax. Plus using `markdown-preview-plus.cson` is most likely a name to never be used by any other package).

Also, the typical workflow I use to hack atom packages is to open a terminal in the repositories directory and then run:
````bash
apm link --dev
````
This adds a system link to `~/.atom/dev/packages`. Then you can load an Atom development instance/window from a terminal with:
````bash
atom --dev
````
which will open the current directory in an Atom development window. A development window will always first look in `~/.atom/dev/packages` for packages to load, so in particular linking your clone of MPP will override MPP installed through Atom, but ONLY when you open a development window. The advantage of this is that if you make a change to your fork that causes Atom to crash, you can still open a normal Atom window that will load the unmodified package that doesn't cause it to crash. I hope that makes sense :D 

Once your done developing open a terminal in the repositories directory and you can unlink it with:
````bash
apm unlink --dev
````

I really appreciate the pull request. No pressure, but if you wanted to implement those changes I would be more than happy to merge your pull request (Im really sorry for the long comment! :D)

KCErb commented:
Those are great suggestions and I'll definitely learn how to implement them and implement them; I knew there was a better way!

Thanks for explaining how to use `link --dev` to work on a package like this, it helps a ton!

KCErb commented:
Woohoo! That was a lot of learning!

This works for me locally now. Some questions:

1.  I went with the name 'latex-macros.cson' during development since it's shorter than `markdown-preview-plus.cson` but it's very simple to change it over . . . unless I don't need to :D. How do you feel about 'latex-macros.cson'? In theory all latex packages would have to use the same syntax right? I'm not too attached though, and `markdown-preview-plus.cson` is definitely the more kosher name so I'm fine with changing it.

2. How should we go about telling users about the feature? Should I add a section to the readme telling them how to name and where to place the file? Or should we include a template file like:
```cson
# You can put your Latex macros here. 
# For example, if you want a boldfaced R to be aliased as \RR then you'd uncomment
# the following:

# RR: "{\\bf R}"

# If you want to have multiple arguments use an array. The following takes 1 argument
# and makes it bold. You use it as \bold{some bold math}

# bold: ["{\\bf #1}",1]
```
If so how do we avoid overwriting the template when updating the package?

3. Do we want to include 'watching'? Meaning something that looks for changes in the macros.cson and updates the MathJax config when they occur? I wonder if MathJax supports that.

4. I'm not a frequent javascript or coffeescript user so I wasn't sure about putting the helper functions (like `loadMacrosFile` and `getUserMacrosPath`) into `module.exports` versus using them directly. The way I wrote them reflects my comfort zone, but if you recommend moving them to `exports` then that seems straightforward enough.

One last note, the biggest challenge I ran into while writing this was finding out that the file methods of `fs-plus` are all asynchronous. First it took me time to realize what was happening but then it took me some time to decide what to do. Ultimately I found out about the synchronous versions of the functions and used those. Does that sound right to you?

Galadirith commented:
@KCErb I'm sorry for taking a week to get back to you, some great work here :D So let me try and answer some of your questions.

1.  **Macro file name**
    I actually think `latex-macros.cson` is a good name and actually describes what its doing. That said I think it should be kept as `markdown-preview-plus.cson` actually for the reason you gave; this syntax actually is specific to MathJax, and other package's that do things with LaTeX I wouldn't expect to understand the files structure. This way it just separates the macro file from confusion with other modules.

2.  **How to inform users**
    Yeh definitely would be great to add some info to the README. What I would do is add a brief comment in the LaTeX feature bullet of the [New Features](https://github.com/Galadirith/markdown-preview-plus/blob/377ff1bdc2412f192ffed6ca2a26e76ee447a549/README.md#new-features) section saying you can define macros in a config file, and then add the full details to [LATEX](https://github.com/Galadirith/markdown-preview-plus/blob/377ff1bdc2412f192ffed6ca2a26e76ee447a549/LATEX.md) by adding a new subsection on Macros and a description of a template file would be a great idea :D.

    Infact we could go one step further and actually create a template file for them someone clicks on a menu option to edit macro's for example. That isn't necessarily something that needs to be done in this pull unless you'd like to :D

3.  **Prevent overwriting when updating MPP**
    So for this we can place the config file in `~/.atom`. You can get that path with `atom.getConfigDirPath()`, and it won't be overwritten when updating.

4.  **Realtime updates**
    So this would defintly be a great feature and other packages like [snippets](https://github.com/atom/snippets/blob/4cf18c7883386eb26e003c4559dd666c92d52739/lib/snippets.coffee#L23) use [pathwatcher](https://github.com/atom/node-pathwatcher). If you want to have a go at this that would be great, and I think yes it could be done but I think it could be easily as complicated as the rest of the work combined. Among the issues is is that you can also define macros in the body of the document with [`\def\<macro-name>{<macro-body>}`](http://docs.mathjax.org/en/latest/tex.html#defining-tex-macros) for example, so how do you determine if when you delete a macro from file, if it should be removed from the running MathJax.

5.  **Module structure**
    The way you've layed out the methods is absolutely fine, infacts its much better not to have them exported, no methods outside of `mathjax-helper` should be calling them :D

6.  **fs-plus and syncronized call's**
    Again I think the way you've implemented the call's synchronously is absolutely fine. We could async it in the furture (I know node dev's love asyncing stuff :D) but unless someones got a massive macro file its not really going to be a problem here :D

7.  **fs.isFileSync call**
    So I could absolutely be missing the point but the call to [`isFileSync`](https://github.com/KCErb/markdown-preview-plus/blob/6376270994743a59df49b170cb41dcaa2ac3918d/lib/mathjax-helper.coffee#L65) needed? [`isObjectPath`](https://github.com/KCErb/markdown-preview-plus/blob/6376270994743a59df49b170cb41dcaa2ac3918d/lib/mathjax-helper.coffee#L56) seems to already check to see if the file exists.
    
8.  **Validating the config file**
    What would be really good is to have some validation of the config file also. So right now whatever is in the config file, if it is successfully loaded as a `cson` object then it will be passed to MathJax.

    Checking that it is actually of the right form before passing to MathJax would be a good idea. So something a simple as iterating over all the top level elements in the object returned by [`loadMacrosFile`](https://github.com/KCErb/markdown-preview-plus/blob/6376270994743a59df49b170cb41dcaa2ac3918d/lib/mathjax-helper.coffee#L55) and making sure that each of the them are of one of the forms:
    - `<macro>: string<body>`
    - `<macro>: [string<body>, int<args>]`
    
9.  **Adding a contributors section to README**
    If you wanted to, please go ahead an add a contributors section to README and add you name to it :D

So ok :D I'll try to distill all the rambling into something coherent. If you could fix/answer questions (1), (2), (3), (7), (8) that would be great. Then we can go for a merge and I'll refactor `mathjax-helper` as a class, which should make it easier to deal with (4) and (6).

Take as long as you need :D and I really need to apologise in advance. I'm going to be super busy over the next couple of months so If I disappear for a few weeks any time over the next couple of months I'm really sorry (why is the not more time in day! :D).

Thanks again @KCErb, really awesome stuff and it will be great to merge in soon, any questions please let me know :D

KCErb commented:
That all sounds great I will do it! I, like you, am super busy but hope to get time in a day or two after this next homework assignment is in!

KCErb commented:
Hey there I finally am back to working on this and I'm really struggling! I'm not very familiar with node and npm!

I updated (`$ git pull upstream master`) my dev fork of MPP so that I could get this working again and even though MPP 1.2.0 is working fine when atom is *not* in dev mode, if fails in dev mode.

After fishing around, it occurred to me that the the node-modules might need to be updated via `npm`. I ran `$ npm update` and that did a lot of something and it *did* clear out the `Expected 41, got 17` errors, but the new error is `Module did not self-register`. The modules that reported this error were `onigurama` `pathwatcher` and `runas`.

I read around a little more and tried `$ npm uninstall onigurama` followed by `$ npm install onigurama` and now I get **two** `Module did not self-register` errors for `onigurama` and just one of the others!!

So I'm now ready to ask for help :)

I'm guessing the registration stuff has something to do with the difference between `apm update` and `npm update`, the only trouble is `apm update` was unresponsive on my dev fork . . .

Thanks for any help you can offer, I'd love to get this back off the ground :)

KCErb commented:
Well I feel dumb :blush: I didn't notice before the suggestion to run `apm rebuild` in the console!!!

That fixed the problem!

Galadirith commented:
@KCErb I'm sorry I wasn't able to offer any advice, but really glad you were able to fix it yourself :D Yeh `apm rebuild` has caught me out a lot in the past too. I really don't understand why `apm install` or `apm update` may sometimes fail but a `npm install` followed by a `apm rebuild` works, but I'm glad it does :D 

If there are any more problems with the node workflow please don't hesitate to post, I have definitely spent many hours ripping my hair out to try to fix such problems in the past (and expect plenty more :D).

KCErb commented:
Ok, this seems to work for me :tada: !

Please let me know what you would have changed. I'll point out things I did inline.

KCErb commented on lib/mathjax-helper.coffee:
```diff
@@ -43,15 +44,72 @@ module.exports =
+  templateFile = fs.readFileSync templatePath, 'utf8'
+  fs.writeFileSync filePath, templateFile
+
+checkMacros = (macrosObject) ->
```
A big part of this PR is not only loading the macros but ensuring they are well formed. 

In these lines I go through each one and make sure it matches the LaTeX rules correctly. I tested these pretty carefully and all seems to work as expected. I'm not sure how to add specs for this though . . .

Galadirith replied:
That's great :D and don't worry about the specs, I need to overhaul the specs so will add some relevant specs for this once everything is merged in :D

KCErb commented on lib/mathjax-helper.coffee:
```diff
@@ -43,15 +44,72 @@ module.exports =
+  for name, value of macrosObject
+    unless name.match(namePattern) and "#{value}".match(valuePattern)
+      delete macrosObject[name]
+      atom.notifications?.addError("Failed to load LaTeX macro named '#{name}'. Please see the [LaTeX guide](https://github.com/Galadirith/markdown-preview-plus/blob/master/LATEX.md#macro-names)", {dismissable: true})
```
This links to my edits on `LaTeX.md`, I'm pretty sure I got the link right, but it's hard to preview until this is published.

KCErb commented on lib/mathjax-helper.coffee:
```diff
@@ -43,15 +44,72 @@ module.exports =
+
+loadUserMacros = ->
+  userMacrosPath = getUserMacrosPath()
+  if fs.isFileSync(userMacrosPath)
```
I ended up keeping this `isFileSync` because it's a convenient place to include generation of the template file. I tried putting it in the `CSON.isObjectPath` but thought a file check was more general and readable . . . agree?

Galadirith replied:
That's fine :D

KCErb commented on lib/mathjax-helper.coffee:
```diff
@@ -43,15 +44,72 @@ module.exports =
+               {.*}    # must be between brackets
+               (,\d)?  # and may be followed by a comma and a digit
+               $///
+
```
These regexs work like I expect them too, but I don't pretend to be a regex expert. If something seems amiss please let me know.

KCErb commented on LATEX.md:
`````diff
@@ -50,3 +50,84 @@ literally display a dollar sign you can use `\$`.
     pull your underwear over your head ...
     ````
+
+## Macros
`````
Is this a tad lengthy? Or OK?

Galadirith replied:
I think its a little lengthy. Would would be great is to distil it down to three main things
- A sentence telling the user that MPP supports persistent macro definitions.
- What file the definitions need to be placed in.
- How to make those definitions in the file (ie the syntax).

Galadirith commented:
@KCErb I just wanted to ping you to say such great work :D I'll hopefully be able to review it early next week, I'm sorry that I'm not able to do it sooner. Thanks again @KCErb

KCErb commented:
Just updated this.

I found out that the brackets are optional, so I changed the way the syntax checker does its work. I also have it checking the arrays more appropriately (in my opinion).

No rush :)

KCErb commented:
Hey @Galadirith! It's been a busy semester, just wondering how things are going and if you need a little `ping` :)

Galadirith commented:
@KCErb How is everything going (and thanks for the ping :D). I'm really sorry that I've been absent for so long, Uni has kept me very busy, among other things.

I am not likely to be able to get round to merging this before June 15th (thats my last exam) which is very unfair on you and I really apologise for that. But I do want to make it absolutely clear that getting your awesome code incorporated is my top priority over any other enhancements.

I'm going to be relativity quite until June 15th, and I'll probably only pop up if theirs some show-stopping bugs around that need squashing :D

KCErb commented:
Hey that sounds fine to me :). I used MPP for all of my notes last semester and it was fantastic. Thanks for all of your dedication to what I feel is a very important project :+1:

Galadirith commented:
@KCErb How's everything going :D I'm sorry i've been away for so long but am ready to get back to work.

Everything looks pretty good and almost ready for a merge (really exciting :D) but there are a couple of changes that I would be grateful if you could make.

The first is that you've required season in [`mathjax-helper.coffee`](https://github.com/Galadirith/markdown-preview-plus/pull/14/files#diff-c1dd0615f39950492816a761e2d6d30fR9) which should also be added to the dependency list in [`package.json`](https://github.com/KCErb/markdown-preview-plus/blob/de523a8ffdd11e86d19a6c1df216a8c236998f43/package.json).

The second is some modifications to [`LATEX.md`](https://github.com/Galadirith/markdown-preview-plus/pull/14/files#diff-b84bfc654444ab62df4fa595c70defd6) and i'll be adding come comment directly to that file.

Thanks so much for your work on this @KCErb and once those changes are made I'll merge it in :D

Galadirith commented on LATEX.md:
```diff
@@ -50,3 +50,75 @@ literally display a dollar sign you can use `\$`.
+This file includes some example macros and explains briefly how to write them.
+If you want to learn even more about them, go to the source: the [MathJax docs](http://docs.mathjax.org/en/latest/tex.html#defining-tex-macros).
+
+## Macro Definition Syntax
```
I think that everything from this point on can be left as it is :D

Galadirith commented on lib/mathjax-helper.coffee:
```diff
@@ -18,7 +19,7 @@ module.exports =
     script.type   = "text/javascript";
     try
-      # atom.packages.resolvePackagePath('mathjax-wrapper') doesnt work but
+      # atom.packages.resolvePackagePath('mathjax-wrapper') doesn't work but
```
Thanks for all the grammar and spelling corrections, I am not good at English :D

KCErb commented:
Hey @Galadirith! I got this but it fell through the cracks, I expect to have some odds and ends time this week to work on this.

Galadirith commented:
No Problem @KCErb :D whenever you've got the time that would be great.

KCErb commented:
Hy @Galadirith finally made those tiny changes you suggested. When I open atom in dev mode it doesn't seem to see the package. So I haven't tested to see if it still works but my only changes are to the LATEX.md and the package.json so I guess it still works.

Sorry, I'm just not sure what has changed in the atom package development world since I last did this so I'm a little lost . . . 

Let me know what else I can do or what I might have overlooked.

leipert commented:
Seems fine to me. @Galadirith Do we want to switch to markdown-it first or do we want to merge this first?

I would really love to publish a better, stable version as 2.0.0 with this PR, markdown-it and maybe scrolling/jumping support? If we publish a new version every week, maybe the users get tired of updating and reading changelogs?

Galadirith commented:
Thanks @KCErb it looks great.

@leipert Yeh I think merge for 2.0.0 makes the most sense. I want to make sure the MathJax component of MPP is as stable as possible for 2.0.0, which must mean adding a good set of specs to it. While doing that I will add specs to cover macro loading by @KCErb's work.

Galadirith commented:
@KCErb Thanks so so much for your hard work on this and your patience. Your codes now in the master branch and will be including in the next release of MPP that will come out this weekend :D

Thanks again @KCErb
